File: 02-integration-clients/src/jira/adf.ts
export interface AdfNode {
  type: string
  text?: string
  content?: AdfNode[]
  attrs?: Record<string, unknown>
  version?: number
}

/**
 * Converts plain text to an ADF doc node.
 * Each non-empty, non-whitespace-only line becomes a separate paragraph.
 * Empty or whitespace-only lines are skipped.
 */
export function toADF(text: string): AdfNode {
  const paragraphs = text
    .split('\n')
    .map((line) => line.trim())
    .filter((line) => line.length > 0)
    .map((line) => ({
      type: 'paragraph',
      content: [{ type: 'text', text: line }],
    }))

  return { version: 1, type: 'doc', content: paragraphs }
}

/**
 * Recursively extracts plain text from an ADF node tree.
 * Returns "" for null/undefined input.
 */
export function adfToText(node: AdfNode | null | undefined): string {
  if (node == null) return ''

  switch (node.type) {
    case 'text':
      return node.text ?? ''

    case 'hardBreak':
      return '\n'

    case 'mention':
      return `@${node.attrs?.text || 'user'}`

    case 'paragraph':
    case 'heading':
    case 'blockquote':
    case 'listItem':
      return (node.content ?? []).map(adfToText).join('') + '\n'

    case 'bulletList':
    case 'orderedList':
      return (node.content ?? []).map(adfToText).join('')

    case 'codeBlock':
      return (node.content ?? []).map(adfToText).join('') + '\n'

    case 'doc':
      return (node.content ?? []).map(adfToText).join('\n').trim()

    default:
      return (node.content ?? []).map(adfToText).join('')
  }
}

File: src/jira/adf.rs
#![allow(dead_code)]

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Atlassian Document Format node, used for Jira descriptions and comments.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdfNode {
    #[serde(rename = "type")]
    pub node_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<Vec<AdfNode>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attrs: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version: Option<u32>,
}

impl AdfNode {
    /// Create a plain text node.
    fn text_node(text: impl Into<String>) -> Self {
        AdfNode {
            node_type: "text".to_string(),
            text: Some(text.into()),
            content: None,
            attrs: None,
            version: None,
        }
    }

    /// Create a paragraph node containing children.
    fn paragraph(children: Vec<AdfNode>) -> Self {
        AdfNode {
            node_type: "paragraph".to_string(),
            text: None,
            content: Some(children),
            attrs: None,
            version: None,
        }
    }
}

/// Convert a plain text string into an ADF document node.
///
/// Each non-empty line becomes a paragraph with a single text node.
pub fn to_adf(text: &str) -> AdfNode {
    let paragraphs: Vec<AdfNode> = text
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(|line| AdfNode::paragraph(vec![AdfNode::text_node(line)]))
        .collect();

    AdfNode {
        node_type: "doc".to_string(),
        text: None,
        content: Some(paragraphs),
        attrs: None,
        version: Some(1),
    }
}

/// Recursively convert an ADF node tree to plain text.
pub fn adf_to_text(node: Option<&AdfNode>) -> String {
    let node = match node {
        Some(n) => n,
        None => return String::new(),
    };

    match node.node_type.as_str() {
        "text" => node.text.clone().unwrap_or_default(),

        "hardBreak" => "\n".to_string(),

        "mention" => {
            // attrs.text or attrs.id
            if let Some(Value::Object(ref attrs)) = node.attrs {
                if let Some(Value::String(name)) = attrs.get("text") {
                    return format!("@{}", name.trim_start_matches('@'));
                }
                if let Some(Value::String(id)) = attrs.get("id") {
                    return format!("@{}", id);
                }
            }
            "@mention".to_string()
        }

        "paragraph" | "heading" | "blockquote" | "listItem" => {
            let children = children_text(node);
            if children.is_empty() {
                "\n".to_string()
            } else {
                format!("{}\n", children)
            }
        }

        "bulletList" | "orderedList" | "codeBlock" => children_text(node),

        "doc" => {
            let raw = children_text(node);
            raw.trim().to_string()
        }

        _ => children_text(node),
    }
}

/// Concatenate plain-text representations of all children.
fn children_text(node: &AdfNode) -> String {
    node.content
        .as_deref()
        .unwrap_or(&[])
        .iter()
        .map(|child| adf_to_text(Some(child)))
        .collect()
}
